### ISSUE TITLE
E2E test is broken

### ISSUE THREAD

**Issue by youngbupark** [dapr/dapr#1060]
## In what area(s)?

/area runtime

## Expected Behavior

service invocation - E2E test should not be broken

## Actual Behavior

service invocation - E2E test is broken.

https://github.com/dapr/dapr/runs/453280944?check_suite_focus=true

```
--- FAIL: TestServiceInvocation (254.60s)
    --- FAIL: TestServiceInvocation/Test_singlehop_for_callee-0 (0.16s)
        service_invocation_test.go:115: 
            	Error Trace:	service_invocation_test.go:115
            	Error:      	Not equal: 
            	            	expected: "singlehop is called"
            	            	actual  : "invalid character 'B' after top-level value"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1 +1 @@
            	            	-singlehop is called
            	            	+invalid character 'B' after top-level value
            	Test:       	TestServiceInvocation/Test_singlehop_for_callee-0
```

## Steps to Reproduce the Problem

Run e2e test

**Comment by youngbupark** [comment]
@jjcollinge looks like metric pr broke the service invocation e2e tests. when we revert your change, it works as expected. I am assuming that the middleware change for grpc might have a bug. 

distributed-calculated add app also thrown invalid header error. 

cc/ @yaron2 @amanbha

**Comment by yaron2** [comment]
Its our bad, we didn't run e2e tests for the PR.

**Comment by yaron2** [comment]
If you run the distributed calculator sample and try to invoke the `add` service with:

`http://localhost:3500/v1.0/invoke/addapp/method/add` 
Method: `POST` 
Sample Body: 

```json
{
	"operandOne": "8",
	"operandTwo": "6"
}
```

You'll get a `400 Bad Request: invalid header value` response. It only happens with the add service.

This looks like its coming from `net/http`: https://golang.org/src/net/http/server.go?s=51112:51172
(Line 997).

That shouldn't be in the hot path, I assume?

**Comment by youngbupark** [comment]
@jjcollinge you can reproduce with [hello-world](https://github.com/dapr/samples/tree/master/1.hello-world) sample. 

build daprd in your machine and replace daprd with this build
```
cd dapr
make build
cp dist/darwin_amd64/daprd /usr/local/bin
dapr --version
```

Run sample hello-world.
```
npm install
dapr run --app-id nodeapp --app-port 3000 --port 3500 node app.js
dapr invoke --app-id nodeapp --method neworder --payload '{"data": { "orderId": "41" } }'
```
dapr invoke command cannot call neworder method

Run python
```
dapr run --app-id pythonapp python3 app.py
```
9090 metric endpoint is conflicting with nodeapp daprd.

So we have two problems:
1. service invocation doesn't work
2. 9090 metric endpoint is conflicting in self-host mode

@jjcollinge Can you please revert your change to not block the other PRs and then create new pr with the fix again ?

**Comment by yaron2** [comment]
> @jjcollinge you can reproduce with [hello-world](https://github.com/dapr/samples/tree/master/1.hello-world) sample.
> 
> build daprd in your machine and replace daprd with this build
> 
> ```
> cd dapr
> make build
> cp dist/darwin_amd64/daprd /usr/local/bin
> dapr --version
> ```
> 
> Run sample hello-world.
> 
> ```
> npm install
> dapr run --app-id nodeapp --app-port 3000 --port 3500 node app.js
> dapr invoke --app-id nodeapp --method neworder --payload '{"data": { "orderId": "41" } }'
> ```
> 
> dapr invoke command cannot call neworder method
> 
> Run python
> 
> ```
> dapr run --app-id pythonapp python3 app.py
> ```
> 
> 9090 metric endpoint is conflicting with nodeapp daprd.
> 
> So we have two problems:
> 
> 1. service invocation doesn't work
> 2. 9090 metric endpoint is conflicting in self-host mode
> 
> @jjcollinge Can you please revert your change to not block the other PRs and then create new pr with the fix again ?

I think the two problems are related.

**Comment by jjcollinge** [comment]
I'm not sure I see the same behaviour with the Hello World sample...

On Ubuntu 18.04, branch master (0007760).

```
make build
cp dist/linux_amd64/release/daprd /usr/local/bin/
```
In hello world sample, master (aef9626).

```
$ dapr run --app-id nodeapp --app-port 3000 --port 3500 node app.js
ℹ️  Starting Dapr with id nodeapp. HTTP Port: 3500. gRPC Port: 41355
✅  You're up and running! Both Dapr and your app logs will appear here.

== DAPR == time="2020-02-18T22:37:43Z" level=info msg="starting Dapr Runtime -- version edge -- commit v0.3.0-rc.0-78-g0007760"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="log level set to: info"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="standalone mode configured"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="dapr id: nodeapp"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="mTLS is disabled. Skipping certificate request and tls validation"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="loaded component messagebus (pubsub.redis)"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="loaded component statestore (state.redis)"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="application protocol: http. waiting on port 3000"
== APP == Node App listening on port 3000!
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="application discovered on port 3000"
== DAPR == 2020-02-18 22:37:43.588824 I | redis: connecting to localhost:6379
== DAPR == 2020-02-18 22:37:43.592659 I | redis: connected to localhost:6379 (localAddr: 127.0.0.1:35966, remAddr: 127.0.0.1:6379)
== DAPR == time="2020-02-18T22:37:43Z" level=warning msg="either no actor state store or multiple actor state stores are specified in the configuration, actor stores specified: 0"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="App is subscribed to the following topics: []"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="Initialized service discovery to standalone"
== DAPR == time="2020-02-18T22:37:43Z" level=warning msg="actors: state store must be present to initialize the actor runtime"
== DAPR == time="2020-02-18T22:37:43Z" level=warning msg="failed to init actors: state store does not support transactions which actors require to save state - please see https://github.com/dapr/docs"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="enabled cors  http middleware"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="enabled proxy http middleware"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="enabled metrics http middleware"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="http server is running on port 3500"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="enabled metrics grpc middleware"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="gRPC server is running on port 41355"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="local service entry announced"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="dapr initialized. Status: Running. Init Elapsed 147.3314ms"
== DAPR == time="2020-02-18T22:37:43Z" level=info msg="starting metrics server on port 9090"
== APP == Got a new order! Order ID: 41
== APP == Successfully persisted state.
-------------------------------------------------------------------
dapr invoke --app-id nodeapp --method neworder --payload '{"data": { "orderId": "41" } }'
✅  App invoked successfully
```
The port conflict is an issue too - it's not exposed in the CLI so the default will conflict

**Comment by youngbupark** [comment]
@jjcollinge oh. yeah. looks like hello-world works as expected.

**Comment by youngbupark** [comment]
Thanks to @jjcollinge , we reverted prometheus change and e2e test turned to green. @jjcollinge will bring up prometheus change with the fix later.

**Comment by jjcollinge** [comment]
Ok so the error is actually raised and returned in the goadder itself because the protomajor and protominor header values are invalid. I've check all other headers and they are ok. I'll update once I've found the issue.
